### DISCUSSION TITLE
Issues with EqualValueConstraint

### DISCUSSION THREAD

**Discussion by TheGreatCid** [idaholab/moose#24768] [Q&A General] [answered]
Hi all,

I am attempting to use an **EqualValueConstraint** to impose continuity across a disjointed mesh. Below is the domain I am using.
 Here the refined region is disjointed from the rest of the mesh. I am trying to impose continuity by matching displacements on the boundaries between f the two meshes. The mesh is fully HEX8
![image](https://github.com/idaholab/moose/assets/75846521/1225363f-fafb-4c37-aeb5-924e4fb7717d)
Here is the constraint block for the x displacement on one of the boundaries 
```  
[mortar_x1]
    type = EqualValueConstraint
    primary_boundary = 'b1_r'
    primary_subdomain = ref_domain
    primary_variable = disp_x
    secondary_boundary = 'b1_c'
    secondary_subdomain = coarse_domain
    variable = lambda
    quadrature = SECOND
  []
```
Here 'ref_domain' is the refined domain and 'coarse_domain' is, well, the coarse domain. "b1_r" and "b1_c" are overlapping boundaries between the domains. I define lamba as a nonlinear variable for the lagrange multiplier. I have this block repeated for all displacements and both boundaries.
When I go to run this simulation, it seg faults. Running MOOSE in debug model reveals the following error in MortarData.C 
```
Assertion `ip' failed
Lower dimensional elements should always have an interior parent set when using mortar
```
I believe the implication of this is that I don't have IDs for all of the volumes on my mesh, but I do. I also have labels for all the surfaces. I am assuming I am misinterpreting this. 
Any clarification would be greatly appreciated!

**Comment by GiudGiud** [comment]
Hello

Can you draw the interface on which you are trying to impose the constraint?
Is it between the coarse and fine elements?

Guillaume

**Reply by TheGreatCid** [reply]
Yes, the interfaces are the surfaces between the coarse and fine elements, as shown here.

![interface](https://github.com/idaholab/moose/assets/75846521/66717549-1c64-443a-8777-4b4ef9309eaf)

**Reply by GiudGiud** [reply]
ok it seems MOOSE is expecting these to be born out of local refinement.

I would try again with a unfirom mesh, set a different subdomain in the "fine" area, then refine using the RefineBlockGeneration

**Reply by TheGreatCid** [reply]
Ah but the issue is then the boundaries are handled using whatever interpolation method MOOSE uses for hanging nodes, and I will start to see stress oscillations since I am simulating a plastic material. I am using this way of doing things to try and avoid that. 
Can the hanging nodes born out of local refinement be handled using Lagrange multipliers instead of mesh interpolation?

**Reply by GiudGiud** [reply]
I m not sure. Would a penalty method work for you?

If the mesh are still separated, even though uniform refinement was used. to get the fine one, you should get the same treatment as you were expecting in your current attempt

**Reply by TheGreatCid** [reply]
A penalty method could work, but I may be misinterpreting the syntax on the constraint conditions. I am taking the primary_boundary to be the interface surfaces between the two domains. Then the primary_subdomain and secondary_domain are the lower dimensional blocks generated from the interface surface side sets. So I am using the LowerDBlockFromSidesetGenerator on the interfaces and using the block it outputs as the respective subdomains.  Do you know if that is correct?

**Reply by GiudGiud** [reply]
going to defer to @recuero @dschwen

**Reply by TheGreatCid** [reply]
Another question to tack on when they see this
![image](https://github.com/idaholab/moose/assets/75846521/98085faf-86da-4f5f-8332-0b6eb772d6a0)
Is there some sort of estimation I could use to figure out what <n> is? It is unclear from this

**Reply by GiudGiud** [reply]
sadly no just trial and error unless you are deep deep into the code. If it gives you the derivatives error, increase it (reconfigure then recompile)

**Reply by recuero** [reply]
> Is there some sort of estimation I could use to figure out what is?

If you want to fine tune, think of the nodal variables from the primary surface that will affect the residual computation on an element on the secondary surface and take its maximum estimate. This number will be bigger with a bigger dimension (e.g. 3 vs 2), with a higher order (second Lagrange vs first Lagrange), and with more physics (e.g. thermal and mechanical primal and dual interface variables).

**Reply by recuero** [reply]
> I am taking the primary_boundary to be the interface surfaces between the two domains. 

There should be a primary boundary and a secondary boundary. You seem to have that defined correctly in your input. Also, the way you are defining lower-dimensional domains seems correct to me. You could try reducing the number of `EqualValueConstraint` constraints in your input until you don't get that particular error anymore and take it from there.

**Reply by lindsayad** [reply]
It would be helpful to see your `Mesh` block

**Reply by TheGreatCid** [reply, marked as answer]
I did actually resolve the previous error. I failed to run one of the boundary conditions I was using through 'LowerDBlockFromSidesetGenerator', but I do have a follow up question. The problem now seems to be behaving OK when using a penalty condition, though not ideally due to stress concentrations at the interface boundaries. 
I would like to use EqualValueConstraint; however, I cannot seem to reach convergence in the linear iterations when switching to EqualValueConstraint. I pasted the relevant sections of my input file, just for a single variable at one of the boundaries.
```
[Mesh]
  [refine]
    type = FileMeshGenerator
    file = '../../gold/torsion_split.msh'
  []
  [secondary1]
    type = LowerDBlockFromSidesetGenerator
    input = refine
    sidesets = 'b1_r'
    new_block_id = 101
    new_block_name = "primary1"
  []
  [primary1]
    type = LowerDBlockFromSidesetGenerator
    input = secondary1
    sidesets = 'b1_c'
    new_block_id = 102
    new_block_name = "secondary1"
  []
[Constraints]
  [mortar_x1]
    type = EqualValueConstraint
    secondary_boundary = 'b1_c'
    primary_boundary = 'b1_r'
    secondary_subdomain = '102'
    primary_subdomain = '101'
    primary_variable = disp_x
    secondary_variable = disp_x
    variable = lx1
    quadrature = SECOND
#    delta = 0.1
    correct_edge_dropping = true
  []
[Variables]
  [disp_x]
  []
  [disp_y]
  []
  [disp_z]
  []
  [lx1]
    block = 'secondary1'
  []
```
Would someone be able to take a glance at this and see if it looks correct? I am using HEX8 elements here so the elements are QUAD4 at the interface.

**Reply by lindsayad** [reply]
Can you paste the bad solve history? Is your Jacobian good? What is your `solve_type`?

**Reply by TheGreatCid** [reply]
It fails on the first time step
Here's all the information:
![image](https://github.com/idaholab/moose/assets/75846521/b24b885c-7586-4dd8-a6a9-dabaa9cbf3dc)
So I am using a NEWTON solver

**Reply by TheGreatCid** [reply]
Ok, it took me a second, but I figured out that the Jacobian is in fact singular here. So no wonder its breaking down

**Reply by TheGreatCid** [reply]
@lindsayad Sorry to bother again, but do you have any thoughts on this?

**Reply by lindsayad** [reply]
You might try using a CONSTANT MONOMIAL type for the lagrange multiplier

**Reply by TheGreatCid** [reply]
Hmm, no luck with that, unfortunately. It seems the behavior in the solver is unchanged

**Reply by lindsayad** [reply]
Well it's possible that you've setup an ill-posed problem though your other BCs, lack of BCs, etc. How many singular values do you have?

**Reply by TheGreatCid** [reply]
It could be ill-posed, though the problem works ok with the EqualValuePenaltyConstraint. It looks like I have 3210 singular values

**Reply by TheGreatCid** [reply]
oh and the more useful point, the svd monitor states that 0 of them are nearly zero

**Reply by lindsayad** [reply]
Have you printed the matrix? How does the scaling compare between primal and LM rows?

**Reply by hugary1995** [reply]
He's running a dynamic analysis (there's a mass matrix) so the matrix can't possibly be singular.

**Reply by hugary1995** [reply]
I agree with @lindsayad -- this is due to either a messed-up Jacobian or a very bad scaling.

**Reply by TheGreatCid** [reply]
Got it, I'll keep debugging then

**Reply by lindsayad** [reply]
> He's running a dynamic analysis (there's a mass matrix) so the matrix can't possibly be singular.

The LM portion could be singular or with very disparate scaling, nearly singular

**Reply by lindsayad** [reply]
equal value constraints do not have diagonals in the LM equations and are unstable, technically even for unequal order primal and LM variable orders. However, even if equal orders are used, there shouldn't be singular values. (I just verified that on our `soln-continuity.i` test input)

**Reply by lindsayad** [reply]
Can you try with `correct_edge_dropping = false` ?

**Reply by TheGreatCid** [reply]
No luck there unfortunately 
It's most likely something off with the domain I've constructed

**Reply by TheGreatCid** [reply]
@lindsayad Actually my mistake! Turning off the edge dropping correction is allowing the simulation to run!

**Reply by lindsayad** [reply]
that's good to hear. It seems like there must be some poor scaling in our "correct" edge dropping treatment. @jbadger95 worked on this and I remember him talking about scaling things. Maybe he can chime in here or I don't know whether @recuero has encountered this?

**Reply by recuero** [reply]
Yes, actual edge dropping can cause scaling issues. Condensing the Lagrange multipliers will improve the condition number.

If turning off the edge dropping correction gets the simulation to run, I'd examine the results for accuracy.

**Reply by TheGreatCid** [reply]
The resulting simulation certainly does not look great. Though it looks similar to the results I achieved with the penalty method.

**Reply by recuero** [reply]
In what sense don't they look great?

**Reply by jbadger95** [reply]
The `correct_edge_dropping` option decides how LMs that only partially project onto the opposing surface are handled. The correct treatment only integrates the primary variable against the LM on the portion of the primary face where the secondary face projects (i.e. where the LM is defined). Integration over partial elements can produce bad scaling when only a small sliver of the element is integrated. The incorrect treatment also integrates the primary variable where the LM isn't defined, effectively enforcing a 0 value on that element. 

So I think there's two possibilities here: 1) there is actually poor scaling due to edge dropping or 2) your problem is not well posed and turning setting `correct_edge_dropping = false` essentially adds an extra constraint. Are you sure that rigid body motions are fully constrained in your setup? IIRC equal value constraints were fairly robust to edge dropping, it was mainly mechanical contact where they were a problem. It's been a couple of years since I looked at mortar methods, but I'd guess it's likely option 2.

Bad scaling can be addressed when a dual basis is used for the LM variable; I think @dewenyushu looked at implementing that but I'm not sure how that turned out. 

Hope that helps!

**Reply by recuero** [reply]
> IIRC equal value constraints were fairly robust to edge dropping, it was mainly mechanical contact where they were a problem

Yes. That's my understanding too. We treat it separately in the code.

**Reply by dewenyushu** [reply]
> Bad scaling can be addressed when a dual basis is used for the LM variable; I think @dewenyushu looked at implementing that but I'm not sure how that turned out.

If the issue is indeed from the bad scaling due to dual basis and edge dropping, then it can be greatly mitigated by using variable condensation preconditioner (VCP)

**Reply by recuero** [reply]
>  Though it looks similar to the results I achieved with the penalty method.

Like @jbadger95 mentioned, most likely there is an issue your problem setup. The issue that started this subthread of "small" projections and ensuing ill conditioning should not happen with penalty; at least, not in the same way. Penalty has issues on its own.

**Reply by TheGreatCid** [reply]
@jbadger95 Thank you for the explanation! It is most likely case 2 that you listed as edge dropping, as you described, is not occurring in the simulation. 

@recuero 
>In what sense don't they look great?

I am seeing some interpenetration of the boundaries as well as stresses occurring on the boundaries where it does not make physical sense. 

>Penalty has issues on its own.

That makes sense. If the problem is repeated between both methods then it is an issue with my problem set-up. I need to go back through and check for rigid body motion+other issues

**Comment by TheGreatCid** [comment]
> > Is there some sort of estimation I could use to figure out what is?
> 
> If you want to fine tune, think of the nodal variables from the primary surface that will affect the residual computation on an element on the secondary surface and take its maximum estimate. This number will be bigger with a bigger dimension (e.g. 3 vs 2), with a higher order (second Lagrange vs first Lagrange), and with more physics (e.g. thermal and mechanical primal and dual interface variables).

This makes sense, thank you!

**Comment by jbadger95** [comment]
Sorry, it's been a while since I looked at mortar methods so my explanation above was a bit off (it's coming back to me now). Edge dropping happens when only part of the primary surface has a corresponding secondary element. In the incorrect treatment the LM is integrated over the entire secondary surface and only part of the primary surface; the part of the secondary surface without a primary counterpart is thus effectively constrained to be 0 (the LM constraint term looks something like <lambda, u_1 - u_2> = 0, so on the portion where u_1 isn't defined it becomes <lamda, -u_2> = 0 which effectively constrains u_2 to be 0 on that portion of the element). I do remember this causing oscillations in solutions (you're trying to enforce discontinuous data so it's somewhat expected); I think I have some examples of that kicking around somewhere.

**Reply by recuero** [reply]
The way it is implemented in MOOSE, edge dropping takes place if part of the secondary side (element) does not project, akin to it sliding past the primary surface (this is just a matter of criterion and may be confusing when compared with the literature); if this doesn't happen, there is no need to modify the LM shape functions and the constraint integrations and using or not using the `correct_edge_dropping` option should render the same results for . This is only considered for mechanical contact on `post()`, so it shouldn't affect the results for the case presented in this discussion.

**Reply by TheGreatCid** [reply]
I misunderstood the definition of edge-dropping previously. I agree; here it should not make a difference as the surfaces are not sliding past each other
